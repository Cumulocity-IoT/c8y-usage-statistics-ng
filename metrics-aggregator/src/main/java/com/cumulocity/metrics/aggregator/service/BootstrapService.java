package com.cumulocity.metrics.aggregator.service;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Service;
import com.cumulocity.microservice.subscription.model.MicroserviceSubscriptionAddedEvent;






/**
 * This is a service to fetch subtenants etc
 * 
 * @author Marco Stoffel
 *
 */

@Service
public class BootstrapService {

    @Autowired
    DeviceMetricsAggregationService deviceMetricsAggregationService;
 
    @Autowired
    StartupUtils utils;

    private static final Logger log = LoggerFactory.getLogger(BootstrapService.class);

    @EventListener
	public void initialize(MicroserviceSubscriptionAddedEvent event) {
		log.info("Tenant Sub: " + event.toString());
        utils.setCurrentTenant(event.getCredentials().getTenant());
        utils.getTenants();
        deviceMetricsAggregationService.createDailyDeviceStatistics();
	}

}
