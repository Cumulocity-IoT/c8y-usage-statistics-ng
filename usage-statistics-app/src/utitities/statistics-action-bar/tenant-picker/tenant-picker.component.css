.hide-item {
    display: none;
}

ul {
    list-style-type: none;
}

li {
    margin: 16.18px
}

.extra {
    display: flex;
    gap: 8.18px;
    cursor: pointer;
}

.negative-margin-top {
    margin-top: -4.045px;
}

.tenant-layout {
    display: flex;
    flex-direction: column;
    gap: 4.045px;
}

.btn-style{
    display: flex;
    gap: 4.045px;
    align-items: center;
}

.item-center{
    align-items: center;
}

.custom-label{
    font-size: x-small;
}
.micro-layout{
    display: flex;
    cursor: pointer;
}
.extra-column{
    flex-direction: column;
    gap: 0;
    color: grey;
}

.li-left{
    margin-left: -16.18px;
}

.disable-tenant{
    pointer-events: none;
    color: grey; 
}