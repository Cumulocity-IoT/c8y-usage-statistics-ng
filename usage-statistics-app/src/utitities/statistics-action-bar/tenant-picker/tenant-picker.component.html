<ng-container *ngIf="currentlyActiveTenant">
  <button type="button" title="Active Tenant" class="btn btn-primary btn-style" [class.btn-pending]="isLoading"
    (click)="openModal(template)"><span class="m-r-4 dlt-c8y-icon-layers"></span> {{currentlyActiveTenant}}</button>
</ng-container>

<ng-template #template>
  <div class="modal-header" style="margin-bottom: 24px;">
    <h4 class="modal-title pull-left">{{'Select a tenant' | translate}}</h4>
    <button type="button" class="btn-close close pull-right" aria-label="Close" (click)="modalRef?.hide()">
      <span aria-hidden="true" class="visually-hidden">&times;</span>
    </button>
  </div>
  <div class="modal-body">
    <ng-container *ngIf="tenantList && tenantList.length > 0">
      <div id="search" class="input-group input-group-search" style="margin-bottom: 16.18px 0px">
        <input type="search" class="form-control" placeholder="Searchâ€¦" [(ngModel)]="searchString"
          (input)="searchTenant()" />
        <span class="input-group-addon">
          <i c8yIcon="search" *ngIf="!searchString || searchString.length === 0"></i>
          <i c8yIcon="times" class="text-muted" *ngIf="searchString && searchString.length > 0"
            (click)="searchString = ''; searchTenant()"></i>
        </span>
      </div>

      <div style="max-height: 54vh; overflow: auto; margin-top: 16.18px">
        <ul>
          <li>
            <div (click)="selectedTenant(sourceTenant)" [class.disable-tenant]="!isActiveTenant(getTenant(sourceTenant))">
              <div class="extra">
                <div class="m-r-4 dlt-c8y-icon-communication-internet" [class.text-primary]="isActiveTenant(getTenant(sourceTenant))"></div>
                <div class="tenant-layout">
                  <div class="extra item-center">
                    <div class="negative-margin-top">{{getDomain(sourceTenant)}}</div>
                    <div class="badge badge-system" style="align-self: flex-start;"
                      *ngIf="getSubtenantCount(sourceTenant) > 0">
                      {{tenantHierarchy[sourceTenant]?.length}}
                    </div>
                  </div>
                  <div class="extra extra-column">
                    <div class="custom-label micro-layout" *ngIf="getCompany(sourceTenant)">
                      <div class="m-r-4 dlt-c8y-icon-company"></div>
                      <div>{{getCompany(sourceTenant)}}</div>
                    </div>
                    <div class="custom-label micro-layout" [title] = 'getDomain(sourceTenant)' *ngIf="getDomain(sourceTenant)">
                      <div class="m-r-4 dlt-c8y-icon-layers"></div>
                          <div>{{sourceTenant}}</div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <ul>
              <li *ngFor="let parent of tenantHierarchy[sourceTenant]" class="li-left">
                <div [class.hide-item]="parent.hideItem" (click)="selectedTenant(parent.id)" [class.disable-tenant]="!isActiveTenant(parent)">
                  <div class="extra">
                    <div class="m-r-4 dlt-c8y-icon-communication-internet" [class.text-primary]="isActiveTenant(parent)"></div>
                    <div class="tenant-layout">
                      <div class="extra item-center">
                        <div class="negative-margin-top">{{getDomain(parent.id)}}</div>
                        <div class="badge badge-system" style="align-self: flex-start;"
                          *ngIf="getSubtenantCount(parent.id) > 0">
                          {{getSubtenantCount(parent.id)}}</div>
                      </div>
                      <div class="extra extra-column">
                        <div class="custom-label micro-layout" *ngIf="getCompany(parent.id)">
                          <div class="m-r-4 dlt-c8y-icon-company"></div>
                          <div>{{getCompany(parent.id)}}</div>
                        </div>
                        <div class="custom-label micro-layout" [title] = 'getDomain(parent.id)' *ngIf="getDomain(parent.id)">
                          <div class="m-r-4 dlt-c8y-icon-layers "></div>
                          <div>{{parent.id}}</div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <ul>
                  <li *ngFor="let child of tenantHierarchy[parent.id]" class="li-left">
                    <div [class.hide-item]="child.hideItem" (click)="selectedTenant(child.id)" [class.disable-tenant]="!isActiveTenant(child)">
                      <div class="extra">
                        <div class="m-r-4 dlt-c8y-icon-communication-internet" [class.text-primary]="isActiveTenant(child)"></div>
                        <div class="tenant-layout">
                          <div class="extra item-center">
                            <div class="negative-margin-top">{{getDomain(child.id)}}</div>
                          </div>
                          <div class="extra extra-column">
                            <div class="custom-label micro-layout" *ngIf="getCompany(child.id)">
                              <div class="m-r-4 dlt-c8y-icon-company"></div>
                              <div>{{getCompany(child.id)}}</div>
                            </div>
                            <div class="custom-label micro-layout" [title] = 'getDomain(child.id)' *ngIf="getDomain(child.id)">
                              <div class="m-r-4 dlt-c8y-icon-layers"></div>
                              <div>{{child.id}}</div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </li>
                </ul>
              </li>
            </ul>
          </li>
        </ul>
      </div>
    </ng-container>
  </div>
</ng-template>