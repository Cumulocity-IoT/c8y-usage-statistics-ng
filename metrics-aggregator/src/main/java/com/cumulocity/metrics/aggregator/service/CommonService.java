package com.cumulocity.metrics.aggregator.service;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Service;

import com.cumulocity.metrics.aggregator.model.microservice.Tenants;
import com.cumulocity.metrics.aggregator.model.microservice.Tenants.Tenant;
import com.cumulocity.microservice.subscription.model.MicroserviceSubscriptionAddedEvent;
import com.cumulocity.microservice.subscription.service.MicroserviceSubscriptionsService;
import com.cumulocity.rest.representation.CumulocityMediaType;
import com.cumulocity.sdk.client.RestConnector;






/**
 * This is a service to fetch subtenants etc
 * 
 * @author Marco Stoffel
 *
 */
@Service
public class CommonService {

    @Autowired
    RestConnector restConnector;
    

    @Autowired
	MicroserviceSubscriptionsService subscriptionsService;

    private static final Logger log = LoggerFactory.getLogger(CommonService.class);

    private String currentTenant = ""; 
    private List<String> tenantList;

    public String getCurrentTenant() {
        return currentTenant;
    }

    public List<String> getTenantList() {
        return tenantList;
    }

    @EventListener
	public void initialize(MicroserviceSubscriptionAddedEvent event) {
		log.info("Tenant Sub: " + event.toString());
        this.currentTenant = event.getCredentials().getTenant();
		log.info("CurrentTenant: " + this.currentTenant);

        this.tenantList = subscriptionsService.callForTenant(this.currentTenant, this::getSubTenants).get();
        this.tenantList.add(currentTenant);
        log.info(tenantList.toString());
	}

    private Optional<List<String>> getSubTenants(){
        Tenants tenants = restConnector.get(
                "/tenant/tenants?pageSize=2000&withTotalPages=true&withApps=false",
                CumulocityMediaType.APPLICATION_JSON_TYPE,Tenants.class);
            ArrayList<String> tenantIds = new ArrayList<String>();
            for (Tenant tenant : tenants.getTenants()){
                tenantIds.add(tenant.getId());
            }
            return Optional.of(tenantIds);
    }

}
