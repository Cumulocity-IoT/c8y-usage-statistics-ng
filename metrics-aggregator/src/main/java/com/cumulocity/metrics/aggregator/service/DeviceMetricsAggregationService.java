package com.cumulocity.metrics.aggregator.service;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.time.Instant;
import java.time.Period;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.context.annotation.Configuration;
import org.springframework.scheduling.annotation.EnableScheduling;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

import com.cumulocity.sdk.client.RestConnector;
import com.cumulocity.sdk.client.SDKException;


import com.cumulocity.metrics.aggregator.model.device.DeviceClassConfiguration;
import com.cumulocity.metrics.aggregator.model.device.DeviceStatistics;
import com.cumulocity.metrics.aggregator.model.device.DeviceStatistics.Statistic;
import com.cumulocity.metrics.aggregator.model.device.DeviceStatisticsAggregation;
import com.cumulocity.metrics.aggregator.model.device.DeviceClassConfiguration.DeviceClass;
import com.cumulocity.microservice.subscription.service.MicroserviceSubscriptionsService;
import com.cumulocity.model.option.OptionPK;
import com.cumulocity.rest.representation.CumulocityMediaType;
import com.cumulocity.rest.representation.tenant.OptionRepresentation;
import com.cumulocity.sdk.client.option.TenantOptionApi;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;



/**
 * This is a service to aggregate device an microservices metrics to display in
 * the usage statistics app
 * It will iterate over all subscribed tenants and aggregate the metrics to an
 * overview
 * 
 * @author Marco Stoffel
 *
 */
@Configuration
@EnableScheduling
@Service
public class DeviceMetricsAggregationService {

	private static final Logger log = LoggerFactory.getLogger(DeviceMetricsAggregationService.class);

	private static final String OPTION_CATEGORY_CONFIGURATION = "configuration";
	private static final String OPTION_KEY = "device.statistics.class.details";

	private Map<String, DeviceClassConfiguration> allDeviceClassConfiguration = new HashMap<String, DeviceClassConfiguration>();
	//private Map<Date,DeviceStatisticsAggregation> deviceStatisticsAggregationCurrentMonth = new HashMap<Date,DeviceStatisticsAggregation>();


	DeviceStatisticsAggregation dailyDeviceStatisticsAggregation = new DeviceStatisticsAggregation();
	Instant dailyDeviceStatisticsAggregationLastRun=  Instant.now().minus(Period.ofDays(2));
	

	@Autowired
	RestConnector restConnector;

	@Autowired
	MicroserviceSubscriptionsService subscriptionsService;


	@Autowired
	TenantOptionApi tenantOptionApi;

	@Autowired
	ObjectMapper objectMapper;


	private List<String> tenantList;

	public List<String> getTenantList() {
		return tenantList;
	}

	public void setTenantList(List<String> tenantList) {
		this.tenantList = tenantList;
	}

	private DateFormat df = new SimpleDateFormat("yyyy-MM-dd");

	public Map<String, DeviceStatistics> getDeviceStatisticsOverview(String type, Date statDate) {
		HashMap<String, DeviceStatistics> dsMap = new HashMap<String, DeviceStatistics>();
			for (String currentTenant : this.getTenantList()) {
				log.info("Get Statistics for Tenant: " + currentTenant);
				
				DeviceStatistics dspage = restConnector.get(
						"/tenant/statistics/device/" + currentTenant + "/" + type + "/" + df.format(statDate)
								+ "?pageSize=2000&withTotalPages=true",
						CumulocityMediaType.APPLICATION_JSON_TYPE, DeviceStatistics.class);
				dspage.setDaysInMonth(statDate,type);
				log.debug("Statistics: " + dspage.getStatistics().toString());
				dsMap.put(currentTenant, dspage);
			}
		return dsMap;
	}

	public Map<String, DeviceClassConfiguration> getAllDeviceClassConfiguration(boolean useTenantDeviceClasses) {
		log.info("# Fetching all deviceClassConfigurations, useTenantDeviceClasses: " + useTenantDeviceClasses);
		this.allDeviceClassConfiguration = new HashMap<String, DeviceClassConfiguration>();
		for (String currentTenant : this.getTenantList()) {
				if (useTenantDeviceClasses) {
					this.allDeviceClassConfiguration.put(currentTenant, getDeviceClasseConfiguration(currentTenant));
				} else {
					this.allDeviceClassConfiguration.put(currentTenant, new DeviceClassConfiguration());
				}

		}
		return allDeviceClassConfiguration;
	}

	private DeviceClassConfiguration getDeviceClasseConfiguration(String currentTenant) {
		log.info("Get DeviceClassConfiguration for Tenant: " + currentTenant);
		final OptionPK option = new OptionPK();
		option.setCategory(OPTION_CATEGORY_CONFIGURATION);
		option.setKey(OPTION_KEY);
		OptionRepresentation optionRepresentation = new OptionRepresentation();
		List<DeviceClassConfiguration.DeviceClass> dcl = new ArrayList<DeviceClassConfiguration.DeviceClass>();
		DeviceClassConfiguration deviceClassConfiguration = new DeviceClassConfiguration();
		try {
			optionRepresentation = tenantOptionApi.getOption(option);
			dcl = new ObjectMapper().readValue(
					optionRepresentation.getValue(),
					new TypeReference<List<DeviceClassConfiguration.DeviceClass>>() {
					});
			deviceClassConfiguration.setDeviceClasses(dcl);
		} catch (SDKException | JsonProcessingException e) {
			log.error("Could not get DeviceClassConfiguration form tenant options using default config.");
		}
		return deviceClassConfiguration;
	}

	@Cacheable(value = "deviceCache", key = "#statDate.toString() + '-' + #type + '-' + #includeSubtenants", condition = "#useTenantDeviceClasses == false")
	public DeviceStatisticsAggregation getAggregatedDeviceClassStatistics(String type, Date statDate,
			boolean includeSubtenants, boolean useTenantDeviceClasses) {
		// Get all Device Statistics for all Tenants

		Map<String, DeviceStatistics> deviceStatisticsMap = this.getDeviceStatisticsOverview(type, statDate);
		Map<String, DeviceClassConfiguration> deviceClassConfigurationMap = this
				.getAllDeviceClassConfiguration(useTenantDeviceClasses);
		// Get all device class configurations
		return getAggregatedDevicesPerClass(deviceStatisticsMap, deviceClassConfigurationMap, includeSubtenants, type);
	}



	private DeviceStatisticsAggregation getAggregatedDevicesPerClass(
			Map<String, DeviceStatistics> deviceStatisticsMap,
			Map<String, DeviceClassConfiguration> deviceClassesMap, boolean includeSubtenants, String type) {

		final Boolean incSubtenants = includeSubtenants;
		// Return object to gather statistics
		DeviceStatisticsAggregation deviceStatisticsAggregation = new DeviceStatisticsAggregation();

		// Loop through all tenans deviceStatistics
		Iterator<Entry<String, DeviceStatistics>> dsIterator = deviceStatisticsMap.entrySet().iterator();
		while (dsIterator.hasNext()) {
			Map.Entry<String, DeviceStatistics> pair = (Entry<String, DeviceStatistics>) dsIterator
					.next();
			String tenant = pair.getKey();
			DeviceStatistics ds = pair.getValue();
			DeviceClassConfiguration deviceClassConfigCurrentTenant = deviceClassesMap.get(tenant);
			// Write amount of devices to total
			deviceStatisticsAggregation.addTotalDevicesCount(ds.getStatistics().size());
			// Loop through all device statistics of current tenant
			DeviceStatisticsAggregation.TenantAggregation ta = new DeviceStatisticsAggregation.TenantAggregation();
			if (includeSubtenants) {
				deviceStatisticsAggregation.getTenantAggregation().put(tenant, ta);
			}
			ds.getStatistics().forEach(statOfSingleDevice -> {
				deviceStatisticsAggregation.addTotalMeas(statOfSingleDevice.getCount());
				if (incSubtenants) {
					ta.addToMeas(statOfSingleDevice.getCount());
					ta.setDeviceClasses(deviceClassConfigCurrentTenant);
					ta.setDevicesCount(ds.getStatistics().size());
					updateDeviceClass(ta.getDeviceClasses(), statOfSingleDevice.getCount(), ds.getDaysInMonth());
				}
				updateDeviceClass(deviceStatisticsAggregation.getTotalDeviceClasses(),
						statOfSingleDevice.getCount(), ds.getDaysInMonth());
			});

		}
		return deviceStatisticsAggregation;
	}

	
	private void updateDeviceClass(DeviceClassConfiguration deviceClassConfiguration, int count, int daysInMonth) {
		float avgMea = count / daysInMonth;
		Iterator<DeviceClass> idc = deviceClassConfiguration.getDeviceClasses().iterator();
		while (idc.hasNext()) {
			DeviceClass dc = idc.next();
			if ((dc.getAvgMaxMea() instanceof String && dc.getAvgMaxMea().equals("INFINITY")
					&& avgMea >= Float.valueOf(dc.getAvgMinMea()))
					||
					(dc.getAvgMaxMea() instanceof Integer && avgMea >= Float.valueOf(dc.getAvgMinMea())
							&& avgMea < Float.valueOf((int) dc.getAvgMaxMea()))) {
				dc.incrementCount();
			}
			;
		}
	}


	public DeviceStatisticsAggregation getDailyStatistics(boolean omitCache){
		Instant now = Instant.now();
		long hours = ChronoUnit.HOURS.between(this.dailyDeviceStatisticsAggregationLastRun, now );
		// fetch only when older than 12 hours
		if (omitCache || hours > 24 )	 {
			log.info("Getting Daily stats.");
			createDailyDeviceStatistics();
		}
		return this.dailyDeviceStatisticsAggregation;
	}


	
	public void createDailyDeviceStatistics(){

			this.dailyDeviceStatisticsAggregation = new DeviceStatisticsAggregation();
			Map<String,Integer> devicesDailyAggregation = new HashMap<String,Integer>();
			Calendar cal = Calendar.getInstance();
			
			int dayOfMonth = cal.get(Calendar.DAY_OF_MONTH);
			cal.set(Calendar.HOUR_OF_DAY, 0);
			cal.set(Calendar.MINUTE, 0);
			cal.set(Calendar.SECOND, 0);
			// cal.set(Calendar.MONTH, 1);
			// dayOfMonth = 28;
			
			// get all device meas from start of month
			for (int i = 1; i <= dayOfMonth; i++){
				cal.set(Calendar.DAY_OF_MONTH, i);
				Date statDate = cal.getTime();
				
				log.info("Get daily statistic: "+ statDate.toString());
				//this.deviceStatisticsAggregationCurrentMonth.put(statDate, getAggregatedDeviceClassStatistics("daily", statDate ,  false, false));
				Map<String, DeviceStatistics> devstat = this.getDeviceStatisticsOverview("daily", statDate);
	
				for (var entry : devstat.entrySet()) {
					List<Statistic> ls =entry.getValue().getStatistics();
					for (Statistic st : ls) {
						devicesDailyAggregation.put(
							st.getDeviceId(), 
							devicesDailyAggregation.getOrDefault(st.getDeviceId(), dayOfMonth) +st.getCount()
							);
					}
				}
			}
			
			// Get device classes
			this.dailyDeviceStatisticsAggregation.setTotalDeviceCount(devicesDailyAggregation.size());
			DeviceClassConfiguration dailyDeviceClasses = new DeviceClassConfiguration();
			for (var device : devicesDailyAggregation.entrySet()){
				this.dailyDeviceStatisticsAggregation.addTotalMeas(device.getValue());
				this.updateDeviceClass(dailyDeviceClasses, device.getValue(), dayOfMonth);
			}
			this.dailyDeviceStatisticsAggregation.setTotalDeviceClasses(dailyDeviceClasses);
			this.dailyDeviceStatisticsAggregationLastRun = Instant.now();
			//log.info("Daily Stat: " + this.deviceStatisticsAggregationCurrentMonth.toString());

    }


	@Scheduled(cron = "0 15 0 * * ?")
	public void scheduleDailyStatistics(){
		subscriptionsService.runForEachTenant(()->{
			this.createDailyDeviceStatistics();
		});
		log.info("ScheduledDailyStatistics");
	}
}
