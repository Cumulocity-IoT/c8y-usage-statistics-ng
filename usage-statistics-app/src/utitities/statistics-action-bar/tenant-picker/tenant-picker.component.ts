import { Component, OnInit, TemplateRef } from '@angular/core';
import { CommonService } from '../../../common.service';
import { BsModalService, BsModalRef } from 'ngx-bootstrap/modal';
import { Alert, AlertService, gettext } from '@c8y/ngx-components';



@Component({
  selector: 'tenant-picker',
  templateUrl: './tenant-picker.component.html',
  styleUrls: ['./tenant-picker.component.css']
})
export class TenantPickerComponent implements OnInit {
  modalRef?: BsModalRef;
  tenantList: any = []
  tenantListDict = {}
  tenantHierarchy: any = []
  sourceTenant: string
  currentlyActiveTenant: string
  searchString = ''
  allTenantIdList = []
  allCompaniesList = []
  allDomainList = []
  isLoading = true;

  constructor(
    private commonService: CommonService,
    private modalService: BsModalService,
    private alertService: AlertService
  ) {
    this.commonService.tenantChanged.subscribe((tenantId: string) => {
      this.currentlyActiveTenant = tenantId
    })
  }

  ngOnInit(): void {
    this.loadSubTenantList()
  }

  async loadSubTenantList() {
    try {
      this.sourceTenant = await this.commonService.getSourceTenant()
      console.log('sourceTenant: ' + this.sourceTenant)
      this.currentlyActiveTenant = await this.commonService.getCurrentlyActiveTenant();
      this.tenantList = await this.commonService.getAllSubtenants(this.sourceTenant);
      this.tenantHierarchy = this.getTenantHierarchyAndBuildDictionary();
      this.allTenantIdList = Object.keys(this.tenantListDict)
      this.allCompaniesList = [... new Set(this.tenantList.map(elem => elem.company))]
      this.allDomainList = [... new Set(this.tenantList.map(elem => elem.domain))]
    }
    catch (error) {
      const alert: Alert = {
        text: gettext('Unable to get the  subtenants of the current tenant. Are there any subtenants attached?'),
        type: 'danger',
        detailedData: error.message
      }
      this.alertService.add(alert)
    }
    finally {
      this.isLoading = false;
    }
  }

  selectedTenant(tenantId) {
    this.currentlyActiveTenant = tenantId
    this.commonService.setCurrentlyActiveTenant(tenantId)
    this.modalRef?.hide()
  }

  getTenantHierarchyAndBuildDictionary() {
    const hierarchy = {}
    this.tenantList.forEach(elem => {
      elem['hideItem'] = false;
      this.tenantListDict[elem.id] = elem;
      if (!elem.parent) {
        elem.parent = "EOF"
      }
      if (hierarchy[elem.parent]) {
        hierarchy[elem.parent].push(elem)
      }
      else {
        hierarchy[elem.parent] = [elem]
      }
    })
    return hierarchy;
  }

  searchTenant() {
    const selectedChildTenantIds = [... new Set([
      ...this.allTenantIdList.filter(elem => elem.includes(this.searchString)),
      ...this.allCompaniesList.filter(elem => elem.toLowerCase().includes(this.searchString.toLowerCase())),
      ...this.allDomainList.filter(elem => elem.toLowerCase().includes(this.searchString.toLowerCase()))
    ])]
    const selectedParentTenantIds = selectedChildTenantIds.map(elem => {
      if (this.tenantListDict[elem]) {
        return this.tenantListDict[elem].parent
      }
    }).filter(elem => elem !== undefined)
    const allChosenTenants = [...selectedChildTenantIds, ...selectedParentTenantIds]
    this.tenantHierarchy[this.sourceTenant].forEach(tenant => {
      tenant['hideItem'] = !(allChosenTenants.includes(tenant.id) || allChosenTenants.includes(tenant.company) || allChosenTenants.includes(tenant.domain));
      if (this.tenantHierarchy[tenant.id]) {
        this.tenantHierarchy[tenant.id].forEach(subTenant => {
          subTenant['hideItem'] = !(allChosenTenants.includes(subTenant.id) || allChosenTenants.includes(subTenant.company) || allChosenTenants.includes(tenant.domain));
          if (!subTenant['hideItem']) {
            const parent = this.tenantHierarchy[this.sourceTenant].find(elem => elem.id === subTenant.parent)
            parent['hideItem'] = false
          }
        })
      }
    })
  }

  openModal(template: TemplateRef<any>) {
    this.modalRef = this.modalService.show(template);
    this.loadSubTenantList()
  }

  getCompany(tenantId: string) {
    return this.tenantListDict[tenantId]?.company
  }
  getDomain(tenantId: string) {
    return this.tenantListDict[tenantId]?.domain
  }
  getSubtenantCount(tenantId: string) {
    return this.tenantHierarchy[tenantId]?.length ?? 0
  }
  isActiveTenant(tenant){
    return tenant?.status === "ACTIVE"
  }
  getTenant(tenantId){
    return this.tenantList.find(elem => elem.id === tenantId)
  }

}