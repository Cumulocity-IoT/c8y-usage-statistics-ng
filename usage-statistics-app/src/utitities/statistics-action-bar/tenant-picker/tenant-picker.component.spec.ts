import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, fakeAsync, TestBed, tick } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { AlertService, CoreModule, ModalService } from '@c8y/ngx-components';
import { Subject } from 'rxjs';
import { TenantPickerComponent } from './tenant-picker.component';
import { CommonService } from '../../../common.service';
import { By } from '@angular/platform-browser';

describe('Test MonthPickerComponent', () => {
  let component: TenantPickerComponent;
  let fixture: ComponentFixture<TenantPickerComponent>;
  let commonService: CommonService;
  let modalService: ModalService;
  let alertService: AlertService;

  const monthPickerServiceMock = {
    selectedDate: new Date("2023-01-01"),
    daysInMonth: 31,
    dateChanged: new Subject()
  }

  const commonServiceMock = {
    tenantChanged:{
        subscribe: jest.fn()
    },
    getSourceTenant: jest.fn(),
    getCurrentlyActiveTenant: jest.fn(),
    getAllSubtenants: jest.fn(),
  }

  const modalServiceMock = {

  }
  
  const alertServiceMock = {
    add: jest.fn()
  }


  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [RouterTestingModule, CoreModule.forRoot()],
      declarations: [TenantPickerComponent],
      schemas: [NO_ERRORS_SCHEMA],
      providers: [
        { provide: CommonService, useValue: commonServiceMock },
        { provide: ModalService, useValue: modalServiceMock },
        { provide: AlertService, useValue: alertServiceMock },
    ]
    });
    fixture = TestBed.createComponent(TenantPickerComponent);
    component = fixture.componentInstance;
    commonService = TestBed.inject(CommonService);
    modalService = TestBed.inject(ModalService);
    alertService = TestBed.inject(AlertService);

  });

  test('should be defined', () => {
    expect(fixture).toBeDefined();
  });

  test('should load expected component', () => {
    expect(fixture).toMatchSnapshot();
  });

  describe('loadSubTenantList', () => {
    it('should create the component with isisLoading set to true by default', () => {
      expect(component.isLoading).toBe(true);
    });

    it('sets isLoading to false when all api calls are done', async () => {
      const mockSourceTenant = 'sourceTenant';
      const mockCurrentlyActiveTenant = 'currentlyActiveTenant';
      const mockTenantList = ['tenant1', 'tenant2'];
  
      jest.spyOn(commonService, 'getSourceTenant').mockResolvedValue(mockSourceTenant);
      jest.spyOn(commonService, 'getCurrentlyActiveTenant').mockResolvedValue(mockCurrentlyActiveTenant);
      jest.spyOn(commonService, 'getAllSubtenants').mockResolvedValue(mockTenantList);
  
      await component.loadSubTenantList();
  
      expect(component.isLoading).toBe(false);
    });
  
    it('should set isLoading to false when an error occurs', async () => {
      const errorMessage = 'Unable to get the  subtenants of the current tenant';

      jest.spyOn(commonService, 'getSourceTenant').mockRejectedValue(() => Promise.reject(new Error(errorMessage)));
  
      await component.loadSubTenantList();
  
      expect(component.isLoading).toBe(false);
      expect(alertService.add).toHaveBeenCalled();
    });

    it('loader on the button should be active if loading is not yet complete', () => {
      component.isLoading = true;
      component.currentlyActiveTenant = 'currentlyActiveTenant';
      fixture.detectChanges();
  
      const button = fixture.debugElement.query(By.css('button.btn.btn-primary'));

      expect(button).toBeTruthy();
      expect(button.nativeElement.classList.contains('btn-pending')).toBeTruthy();
    });
    
  
    it('loader on the button should not be active when loading is complete', () => {
      component.isLoading = false;
      component.currentlyActiveTenant = 'currentlyActiveTenant';
      fixture.detectChanges();
    
      const button = fixture.debugElement.query(By.css('.btn.btn-primary'));

      expect(button).toBeTruthy();
      expect(button.nativeElement.classList.contains('btn-pending')).toBeFalsy();
    });

  });
});