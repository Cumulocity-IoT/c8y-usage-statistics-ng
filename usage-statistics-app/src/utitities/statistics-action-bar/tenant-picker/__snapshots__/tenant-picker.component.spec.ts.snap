// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test MonthPickerComponent should load expected component 1`] = `
<tenant-picker
  alertService={[Function Object]}
  allCompaniesList={[Function Array]}
  allDomainList={[Function Array]}
  allTenantIdList={[Function Array]}
  commonService={[Function Object]}
  isLoading={[Function Boolean]}
  modalService={[Function BsModalService]}
  searchString=""
  tenantHierarchy={[Function Array]}
  tenantList={[Function Array]}
  tenantListDict={[Function Object]}
>
</tenant-picker>
`;
